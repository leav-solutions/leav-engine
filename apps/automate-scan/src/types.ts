// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {Channel} from 'amqplib';

export interface IAmqpParams {
    channel?: Channel;
    exchange?: string;
    routingKey?: string;
}

export interface IWatcherParams {
    awaitWriteFinish?: {
        stabilityThreshold?: number;
        pollInterval: 100;
    };
    delay?: number;
    verbose?: boolean | 'very';
}

export interface IConfig {
    allowFilesList: string;
    ignoreFilesList: string;
    rootPath: string;
    rootKey?: string;
    redis: {
        host: string;
        port: number;
    };
    amqp?: {
        protocol: string;
        hostname: string;
        port: number;
        username: string;
        password: string;
        queue: string;
        exchange: string;
        routingKey: string;
        type: string;
    };
    watcher?: {
        awaitWriteFinish: {
            stabilityThreshold: number;
            pollInterval: number;
        };
    };
    verbose?: IWatcherParams['verbose'];
}

export interface IParams {
    rootPath: string;
    rootKey: string;
    verbose: IWatcherParams['verbose'];
    amqp?: IAmqpParams;
}

export interface IParamsExtends extends IParams {
    delay?: number;
    ready?: boolean;
}

export interface IMessageSend {
    event: string;
    time: number;
    pathAfter: string;
    pathBefore: string;
    inode: number;
    isDirectory: boolean;
    rootKey: any;
    hash: string;
}
