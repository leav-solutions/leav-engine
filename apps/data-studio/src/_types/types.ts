// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt

import {GET_ATTRIBUTES_BY_LIB_attributes_list_StandardAttribute_embedded_fields} from '_gqlTypes/GET_ATTRIBUTES_BY_LIB';
import {AttributeFormat, AttributeType, RecordFilterCondition} from '_gqlTypes/globalTypes';
import {RecordIdentity_whoAmI} from '_gqlTypes/RecordIdentity';
import {
    ILibraryDetailExtendedAttributeParentLinkedLibrary,
    ILibraryDetailExtendedAttributeParentLinkedTree
} from '../graphQL/queries/libraries/getLibraryDetailExtendQuery';

export interface ISystemTranslation {
    [lang: string]: string;
}

export interface ILibrary {
    id: string;
    label: ISystemTranslation;
}

export interface IItemBase {
    fields: {[x: string]: any};
    whoAmI: RecordIdentity_whoAmI;
    index: number;
}

export type IItem = IItemBase;

export interface IRecordIdentityWhoAmI extends RecordIdentity_whoAmI {
    [x: string]: any;
    index?: number;
}

export enum PreviewSize {
    tiny = 'tiny',
    small = 'small',
    medium = 'medium',
    big = 'big'
}

export enum AvailableLanguage {
    en = 'en',
    fr = 'fr'
}

export enum FilterType {
    ATTRIBUTE = 'ATTRIBUTE',
    TREE = 'TREE',
    LIBRARY = 'LIBRARY'
}

export interface IFilter {
    type: FilterType;
    index: number; // use to sort the filters
    key: string; // attribute / tree key
    value: {value: boolean | string | number | null | IDateRangeValue; label?: string};
    active: boolean;
    condition: AttributeConditionType | TreeConditionFilter | ThroughConditionFilter;
}

export enum OperatorFilter {
    AND = 'AND',
    OR = 'OR',
    OPEN_BRACKET = 'OPEN_BRACKET',
    CLOSE_BRACKET = 'CLOSE_BRACKET'
}

export enum TreeConditionFilter {
    CLASSIFIED_IN = 'CLASSIFIED_IN',
    NOT_CLASSIFIED_IN = 'NOT_CLASSIFIED_IN'
}

export enum ThroughConditionFilter {
    THROUGH = 'THROUGH'
}

// We're exporting a const and not an enum to "merge" the enum coming
// from Graphql types with some condition of our own
export const AttributeConditionFilter = {
    ...RecordFilterCondition,
    ...ThroughConditionFilter
};

export type AttributeConditionType = ValueOf<typeof AttributeConditionFilter>;

export enum OrderSearch {
    DESC = 'DESC',
    ASC = 'ASC'
}

export interface IAttribute {
    id: string;
    library: string;
    type: AttributeType;
    format?: AttributeFormat;
    label: SystemTranslation | null;
    isLink: boolean;
    isMultiple: boolean;
    linkedLibrary?: ILibraryDetailExtendedAttributeParentLinkedLibrary;
    linkedTree?: ILibraryDetailExtendedAttributeParentLinkedTree;
    parentAttribute?: IAttribute;
    embedded_fields?: Array<GET_ATTRIBUTES_BY_LIB_attributes_list_StandardAttribute_embedded_fields | null> | null;
}

export interface IEmbeddedFieldData {
    path: string;
    format: AttributeFormat;
}

export interface IFieldBase {
    id: string;
    library: string;
    label: string;
    key: string;
    format?: AttributeFormat;
    embeddedData?: IEmbeddedFieldData;
    multipleValues?: boolean;
}

export interface IFieldTypeBasic extends IFieldBase {
    type: AttributeType.simple | AttributeType.advanced;
    parentAttributeData?: IParentAttributeData;
}

export interface IFieldTypeLink extends IFieldBase {
    type: AttributeType.simple_link | AttributeType.advanced_link;
    parentAttributeData?: IParentAttributeData;
}

export interface IFieldTypeTree extends IFieldBase {
    type: AttributeType.tree;
    parentAttributeData?: IParentAttributeData;
    recordLibrary?: string;
    treeData?: ITreeData;
}

export type IField = IFieldTypeBasic | IFieldTypeLink | IFieldTypeTree;

export interface IParentAttributeData {
    id: string;
    type: AttributeType;
}

export interface ITreeData {
    treeAttributeId: string;
    libraryTypeName: string;
}

export interface ITree {
    id: string;
    label: ISystemTranslation | null;
    libraries: Array<{
        library: {
            id: string;
            label: ISystemTranslation | null;
        };
    }>;
}

export interface INavigationPath {
    id: string;
    library: string;
    label?: string | null;
}

export enum InfoType {
    basic = 'basic',
    success = 'success',
    warning = 'warning',
    error = 'error'
}

export enum InfoPriority {
    low = 'low',
    medium = 'medium',
    high = 'high'
}

export enum InfoChannel {
    trigger = 'trigger',
    passive = 'passive'
}

export interface IBaseInfo {
    content: string;
    type: InfoType;
}

export interface IInfo extends IBaseInfo {
    time?: number;
    priority?: InfoPriority;
    channel?: InfoChannel;
}

export interface ISharedStateSelectionSearch {
    type: SharedStateSelectionType.search;
    selected: ISharedSelected[];
    filters?: IFilter[];
    allSelected?: boolean;
}

export interface ISharedStateSelectionNavigation {
    type: SharedStateSelectionType.navigation;
    selected: ISharedSelected[];
    parent: string;
}

export type SharedStateSelection = ISharedStateSelectionSearch | ISharedStateSelectionNavigation;

export enum SharedStateSelectionType {
    navigation,
    search
}

export interface ISharedSelected {
    id: string;
    library: string;
    label: string;
    nodeId?: string;
}

export enum WorkspacePanels {
    HOME = 'home',
    LIBRARY = 'library',
    TREE = 'tree'
}
export interface IDateRangeValue {
    from: string;
    to: string;
}

export type ValueOf<T> = T[keyof T];

export interface IApplicationSettings {
    libraries?: 'all' | 'none' | string[];
    librariesOrder?: string[];
    trees?: 'all' | 'none' | string[];
    treesOrder?: string[];
    showTransparency?: boolean;
}

export enum NotifTypes {
    TASK = 'TASK'
}
