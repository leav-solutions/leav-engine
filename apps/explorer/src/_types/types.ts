// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt

import {GET_ATTRIBUTES_BY_LIB_attributes_list_StandardAttribute_embedded_fields} from '_gqlTypes/GET_ATTRIBUTES_BY_LIB';
import {
    AttributeFormat,
    AttributeType,
    RecordFilterCondition,
    RecordFilterOperator,
    ViewSizes,
    ViewTypes
} from '_gqlTypes/globalTypes';
import {RecordIdentity, RecordIdentity_whoAmI} from '_gqlTypes/RecordIdentity';
import {TREE_NODE_CHILDREN_treeNodeChildren_list_permissions} from '_gqlTypes/TREE_NODE_CHILDREN';
import {
    ILibraryDetailExtendedAttributeParentLinkedLibrary,
    ILibraryDetailExtendedAttributeParentLinkedTree
} from '../graphQL/queries/libraries/getLibraryDetailExtendQuery';
import {IGetViewListSettings, IGetViewListSort} from '../graphQL/queries/views/getViewsListQuery';

export interface ISystemTranslation {
    [lang: string]: string;
}

export interface ILibrary {
    id: string;
    label: ISystemTranslation;
    gqlNames: {
        query: string;
        filter: string;
        searchableFields: string;
    };
}

export enum LinkedType {
    library = 'library',
    tree = 'tree'
}

export type FilePreview = {
    tiny: string;
    small: string;
    medium: string;
    big: string;
    huge: string;
    pages: string;
} | null;

export interface IItemBase {
    fields: {[x: string]: any};
    whoAmI: RecordIdentity_whoAmI;
    index: number;
}

export type IItem = IItemBase;

export interface IRecordIdentityWhoAmI extends RecordIdentity_whoAmI {
    [x: string]: any;
    index?: number;
}

export enum PreviewAttributes {
    tiny = 'tiny',
    small = 'small',
    medium = 'medium',
    big = 'big',
    huge = 'huge',
    pages = 'pages'
}

export enum PreviewSize {
    small = 'small',
    medium = 'medium',
    big = 'big'
}

export enum AvailableLanguage {
    en = 'en',
    fr = 'fr'
}

export enum FilterTypes {
    filter = 'filter',
    separator = 'separator'
}

export interface IFilterSeparatorCommon {
    key: number;
    id: string;
}

export enum FilterType {
    ATTRIBUTE = 'ATTRIBUTE',
    TREE = 'TREE',
    LIBRARY = 'LIBRARY'
}

export interface IFilter {
    type: FilterType;
    index: number; // use to sort the filters
    key: string; // attribute / tree key
    value: {value: boolean | string | number | null | IDateRangeValue; label?: string};
    active: boolean;
    condition: AttributeConditionType | TreeConditionFilter | ThroughConditionFilter;
}

export interface IFilterAttribute extends IFilter {
    attribute: IAttribute; // Put the attribute in the filter to avoid having to fetch them multiple times
    parentTreeLibrary?: IFilterLibrary; // on tree library attribute
}

export interface IFilterTree extends IFilter {
    tree: {id: string; label?: ISystemTranslation | null};
}

// on library's tree
export interface IFilterLibrary extends IFilter {
    library: {id: string; label?: ISystemTranslation | null};
    parentAttribute: IAttribute;
}

export enum OperatorFilter {
    AND = 'AND',
    OR = 'OR',
    OPEN_BRACKET = 'OPEN_BRACKET',
    CLOSE_BRACKET = 'CLOSE_BRACKET'
}

export enum TreeConditionFilter {
    CLASSIFIED_IN = 'CLASSIFIED_IN',
    NOT_CLASSIFIED_IN = 'NOT_CLASSIFIED_IN'
}

export enum ThroughConditionFilter {
    THROUGH = 'THROUGH'
}

// We're exporting a const and not an enum to "merge" the enum coming
// from Graphql types with some condition of our own
export const AttributeConditionFilter = {
    ...RecordFilterCondition,
    ...ThroughConditionFilter
};

export type AttributeConditionType = ValueOf<typeof AttributeConditionFilter>;

export interface IQueryFilter {
    field?: string;
    value?: string;
    condition?: RecordFilterCondition;
    operator?: RecordFilterOperator;
    treeId?: string;
}

export enum OrderSearch {
    DESC = 'DESC',
    ASC = 'ASC'
}

export interface IAttribute {
    id: string;
    library: string;
    type: AttributeType;
    format?: AttributeFormat;
    label: SystemTranslation | null;
    isLink: boolean;
    isMultiple: boolean;
    linkedLibrary?: ILibraryDetailExtendedAttributeParentLinkedLibrary;
    linkedTree?: ILibraryDetailExtendedAttributeParentLinkedTree;
    parentAttribute?: IAttribute;
    embedded_fields?: Array<GET_ATTRIBUTES_BY_LIB_attributes_list_StandardAttribute_embedded_fields | null> | null;
}

export type ExtendFormat = string | {[key: string]: ExtendFormat[]};

export interface IEmbeddedFieldData {
    path: string;
    format: AttributeFormat;
}

export interface IFieldBase {
    id: string;
    library: string;
    label: string;
    key: string;
    format?: AttributeFormat;
    embeddedData?: IEmbeddedFieldData;
    multipleValues?: boolean;
}

export interface IFieldTypeBasic extends IFieldBase {
    type: AttributeType.simple | AttributeType.advanced;
    parentAttributeData?: IParentAttributeData;
}

export interface IFieldTypeLink extends IFieldBase {
    type: AttributeType.simple_link | AttributeType.advanced_link;
    parentAttributeData?: IParentAttributeData;
}

export interface IFieldTypeTree extends IFieldBase {
    type: AttributeType.tree;
    parentAttributeData?: IParentAttributeData;
    recordLibrary?: string;
    treeData?: ITreeData;
}

export type IField = IFieldTypeBasic | IFieldTypeLink | IFieldTypeTree;

export interface IRecordEdition {
    show: boolean;
    item?: IItem;
}

export interface IAccordionActive {
    id: string;
    library: string;
    depth: number;
}

export interface ISelectedAttribute {
    id: string;
    library: string;
    path: string;
    label: SystemTranslation | null;
    type: AttributeType;
    format?: AttributeFormat | null;
    multiple_values: boolean;
    parentAttributeData?: IParentAttributeData;
    embeddedFieldData?: IEmbeddedFields;
    treeData?: ITreeData;
}

export interface IParentAttributeData {
    id: string;
    type: AttributeType;
}

export interface ITreeData {
    treeAttributeId: string;
    libraryTypeName: string;
}

export interface IEmbeddedFields {
    id: string;
    format: AttributeFormat;
    label: ISystemTranslation;
    embedded_fields: IEmbeddedFields[];
}

export interface IGroupEmbeddedFields {
    [attributeId: string]: {
        embedded_fields: {[key: string]: IEmbeddedFields};
    };
}

export interface IAttributeSelected {
    id: string;
    library: string;
    originAttributeData?: IParentAttributeData;
    extendedData?: IEmbeddedFieldData;
    treeData?: ITreeData;
}

export interface ITree {
    id: string;
    label: ISystemTranslation | null;
    libraries: Array<{
        library: {
            id: string;
            label: ISystemTranslation | null;
        };
    }>;
}

export interface INavigationPath {
    id: string;
    library: string;
    label?: string | null;
}

export enum NotificationType {
    basic = 'basic',
    success = 'success',
    warning = 'warning',
    error = 'error'
}

export enum NotificationPriority {
    low = 'low',
    medium = 'medium',
    high = 'high'
}

export enum NotificationChannel {
    trigger = 'trigger',
    serverError = 'passive'
}

export interface IBaseNotification {
    content: string;
    type: NotificationType;
}

export interface INotification extends IBaseNotification {
    time?: number;
    priority?: NotificationPriority;
    channel?: NotificationChannel;
}

export type ILang = string[];

export enum TypeSideItem {
    filters = 'filters',
    view = 'view'
}

export interface IView {
    id: string;
    library?: string;
    label: ISystemTranslation;
    display: IViewDisplay;
    owner: boolean;
    shared: boolean;
    description?: ISystemTranslation;
    color?: string;
    filters?: IFilter[];
    settings?: IGetViewListSettings[];
    sort?: IGetViewListSort;
}

export interface IViewDisplay {
    type: ViewTypes;
    size: ViewSizes;
}

export interface ILinkedElement {
    id: string;
    linkedType: LinkedType;
}

export interface ITableCell {
    value: any;
    type: AttributeType;
    format?: AttributeFormat;
}

export interface ITableRow {
    record: RecordIdentity_whoAmI;
    [x: string]: ITableCell | IRecordIdentityWhoAmI | string;
}

export interface ISharedStateSelectionSearch {
    type: SharedStateSelectionType.search;
    selected: ISharedSelected[];
    allSelected?: boolean;
}

export interface ISharedStateSelectionNavigation {
    type: SharedStateSelectionType.navigation;
    selected: ISharedSelected[];
    parent: string;
}

export type SharedStateSelection = ISharedStateSelectionSearch | ISharedStateSelectionNavigation;

export enum SharedStateSelectionType {
    navigation,
    search
}

export interface ISharedSelected {
    id: string;
    library: string;
    label: string;
    nodeId?: string;
}

export interface IToggleSelection {
    selectionType: SharedStateSelectionType;
    elementSelected: ISharedSelected;
    parent?: string;
}

export enum WorkspacePanels {
    HOME = 'home',
    LIBRARY = 'library',
    TREE = 'tree'
}
export interface IDateRangeValue {
    from: string;
    to: string;
}

export type ValueOf<T> = T[keyof T];

export interface IPermissions {
    [key: string]: boolean;
}

export interface ITreeContentRecordAndChildren {
    id: string;
    record: RecordIdentity;
    children?: ITreeContentRecordAndChildren[];
    permissions: TREE_NODE_CHILDREN_treeNodeChildren_list_permissions;
}
