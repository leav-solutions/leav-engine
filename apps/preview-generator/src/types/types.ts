// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
export interface IConfig {
    inputRootPath: string;
    outputRootPath: string;
    ICCPath: string;
    amqp: {
        protocol: string;
        hostname: string;
        port: number;
        username: string;
        password: string;
        type: string;
        consume: {
            queue: string;
            exchange: string;
            routingKey: string;
        };
        publish: {
            queue: string;
            exchange: string;
            routingKey: string;
        };
    };
    verbose?: boolean;
}

export interface IResult {
    error: number;
    error_detail?: string;
    params?: {
        background?: true | false | string;
        density?: number;
        size?: number;
        output: string;
        name: string;
    };
}

export interface IResponse {
    context: any;
    input: string;
    results?: IResult[];
}

export interface IMessageConsume {
    input: string;
    context: any;
    versions: IVersion[];
}

export interface IVersion {
    background?: true | false | string;
    density?: number;
    pdf?: string;
    sizes: ISize[];
}

export interface ISize {
    size: number;
    output: string;
    name: string;
}

export interface IExec {
    command: string;
    args: string[];
}

export interface IArgs {
    before: string[];
    after: string[];
}

export interface IRootPaths {
    input: string;
    output: string;
}
