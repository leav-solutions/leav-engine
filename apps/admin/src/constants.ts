// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt

export const APPS_ENDPOINT = 'app';
export const API_ENDPOINT = 'graphql';
export const APP_ENDPOINT = window.location.pathname.split('/').filter(e => e)[1]; // Get endpoint app current from url /APPS_ENDPOINT/:endpoint
export const LOGIN_ENDPOINT = 'login';
export const ORIGIN_URL = window.location.origin;
export const WS_URL = `${window.location.protocol === 'https:' ? 'wss' : 'ws'}://${window.location.host}`;
