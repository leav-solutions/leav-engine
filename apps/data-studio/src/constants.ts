// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt

export const API_ENDPOINT = 'graphql';
export const LOGIN_ENDPOINT = 'login';
export const ORIGIN_URL = window.location.origin;
export const WS_URL = `${window.location.protocol === 'https:' ? 'wss' : 'ws'}://${window.location.host}`;

export const UNAUTHENTICATED = 'UNAUTHENTICATED';
export const PREFIX_USER_VIEWS_ORDER_KEY = 'user_views_order_';
export const PREFIX_SHARED_VIEWS_ORDER_KEY = 'shared_views_order_';

export const FAVORITE_LIBRARIES_KEY = 'favorites_libraries_ids';
export const FAVORITE_TREES_KEY = 'favorites_trees_ids';

export const INFOS_COLUMN_WIDTH = '350px';
