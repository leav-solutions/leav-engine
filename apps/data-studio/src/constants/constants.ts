// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {AttributeFormat, ViewSizes, ViewTypes} from '_gqlTypes/globalTypes';
import {IView} from '../_types/types';

export const selectionColumn = 'selection-column';
export const infosCol = 'infos';

export const defaultInfosTime = 5000;

export const panelSize = '22.5rem';

export const initialColumnsLimit = 5;

export const viewSettingsField = 'fields';

export const attributeExtendedKey = 'extended';

export const defaultView: IView = {
    id: 'default-view',
    label: {fr: 'Vue par dÃ©faut', en: 'Default view'},
    owner: true,
    display: {type: ViewTypes.list, size: ViewSizes.MEDIUM},
    shared: false,
    filters: []
};

export const formatNotUsingCondition = [AttributeFormat.boolean];

export const defaultLinkAttributeFilterFormat = AttributeFormat.text;

export const treeNavigationPageSize = 20;

export const getSelectedViewKey = (libraryId: string) => `selected_view_${libraryId}`;

export const APPS_BASE_URL = import.meta.env.VITE_CORE_URL + '/' + import.meta.env.VITE_ENDPOINT_BASE;
