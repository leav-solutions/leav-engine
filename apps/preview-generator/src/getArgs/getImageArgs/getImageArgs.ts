// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {join} from 'path';
import {ErrorPreview} from '../../errors/ErrorPreview';
import {Colorspaces} from '../../types/constants';
import {IExec, IVersion} from '../../types/types';
import {handleError} from '../../utils/log';
import {getConfig} from './../../getConfig/getConfig';
import {getSvgCommand} from './../getSvgCommand/getSvgCommand';
import {handleBackground} from './handleBackground/handleBackground';
import {getColorspace} from './helpers/getColorspace';
import {hasClippingPath} from './helpers/hasClippingPath';
import {hasTransparency} from './helpers/hasTransparency';

const _getMainCommand = async (
    ext: string,
    input: string,
    output: string,
    size: number,
    name: string,
    version: IVersion,
    first = false
): Promise<IExec> => {
    if (ext === 'svg') {
        return getSvgCommand(input, output, size);
    }

    const inputFile = `${input}[0]`;
    const command = 'magick';
    const config = await getConfig();

    const inputHasClippingPath = await hasClippingPath(inputFile);

    let profileArgs = [];
    let clippingPathArgs: string[] = [];

    // Hardcoded value in input, to process the image with better quality.
    // User defined density (eg. 72 or 300) is used in output
    const densityArgs = [
        '-density', // use density option
        600 // density value
    ];

    const colorspaceArgs = ['-colorspace', 'srgb'];
    let stripArgs = [];

    if (first) {
        const colorspace = await getColorspace(input);

        if (inputHasClippingPath) {
            const inputHasTransparency = await hasTransparency(inputFile);

            // Select the outside of the clipping path, and apply transparency
            clippingPathArgs = ['+clip-path', '#1', '-alpha', 'transparent', '-evaluate', 'divide', '0'];

            // Select the inside of the clipping path, and set it opaque
            if (!inputHasTransparency) {
                clippingPathArgs = [...clippingPathArgs, '-clip', '-alpha', 'opaque'];
            }
        }

        if (colorspace === Colorspaces.CMYK) {
            profileArgs = [
                ...profileArgs,
                '-profile', // use profile option
                join(config.ICCPath, 'eciCMYK_v2.icc') // profile value
            ];
        }

        profileArgs = [
            ...profileArgs,
            '-profile', // use profile option
            join(config.ICCPath, 'eciRGB_v2.icc') // profile value
        ];

        stripArgs = ['-strip'];
    }

    const densityOutArgs = [
        '-density', // use density option
        version.density ? version.density.toString() : '72' // density value
    ];

    const resizeArgs: string[] = [
        '-geometry', // use resize option
        `${size}x${size}>` // resize value
    ];

    const args: string[] = [
        ...densityArgs,
        inputFile,
        ...profileArgs,
        ...colorspaceArgs,
        ...clippingPathArgs,
        ...resizeArgs,
        ...densityOutArgs,
        ...stripArgs,
        `png:${output}` // output path
    ];

    return {command, args};
};

export const getImageArgs = async (
    ext: string,
    input: string,
    output: string,
    size: number,
    name: string,
    version: IVersion,
    first = false
): Promise<IExec[]> => {
    try {
        const mainCommand = await _getMainCommand(ext, input, output, size, name, version, first);

        let backgroundCommand: IExec = null;
        if (first && version.background !== false) {
            backgroundCommand = handleBackground(version.background, output);
        }

        return [mainCommand, backgroundCommand];
    } catch (err) {
        const errorId = handleError(err);

        throw new ErrorPreview({
            error: 504,
            params: {
                background: version.background,
                density: version.density,
                size,
                output,
                name,
                errorId
            }
        });
    }
};
