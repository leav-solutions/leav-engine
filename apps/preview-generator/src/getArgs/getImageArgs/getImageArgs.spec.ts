// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {Colorspaces} from '../../types/constants';
import {IVersion} from '../../types/types';
import {getImageArgs} from './getImageArgs';

jest.mock('./helpers/getColorspace', () => ({
    getColorspace: jest.fn(() => Promise.resolve(Colorspaces.CMYK))
}));

jest.mock('./helpers/hasClippingPath', () => ({
    hasClippingPath: () => Promise.resolve(true)
}));

jest.mock('./helpers/hasTransparency', () => ({
    hasTransparency: () => Promise.resolve(true)
}));

describe('getImageArgs', () => {
    test('Compute preview command args', async () => {
        const ext = 'jpg';
        const input = 'test.jpg';
        const output = 'test.png';
        const size = 800;
        const name = 'medium';

        const version: IVersion = {
            sizes: [{output, size, name}]
        };

        const args = await getImageArgs(ext, input, output, size, name, version);

        // Args might be complicated and subject to change. Check snapshot to see if this looks fine
        expect(args).toMatchSnapshot();
    });
});
