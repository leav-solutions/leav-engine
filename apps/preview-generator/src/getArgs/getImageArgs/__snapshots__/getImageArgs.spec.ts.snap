// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getImageArgs Compute preview command args 1`] = `
[
  {
    "args": [
      "-density",
      600,
      "test.jpg[0]",
      "-colorspace",
      "srgb",
      "-geometry",
      "800x800>",
      "-density",
      "72",
      "png:test.png",
    ],
    "command": "magick",
  },
  null,
]
`;
